### ISSUE TITLE
Inferred return type not correct when using first-class callable

### ISSUE THREAD

**Issue by annuh** [vimeo/psalm#11038]
I found that Psalm (and [PHPStan](https://github.com/phpstan/phpstan/issues/11305)) could not determine the inferred return type correctly when using a First-class Callable. I think it should behave the same as when using a callable via a string. Hopefully the snippet demonstrates the issue🙂

https://psalm.dev/r/b3176fce9b

The expected result is no issues are found.

**Comment by mpesari** [comment]
I run into another problem with first class callable syntax, which seems so related to this that I decided not to create another issue:

https://psalm.dev/r/d0cc7f5046

I think this will be bit of a problem with PHP 8.4 using `array_any()`:

```php
<?php

if (array_any(['foo',1,2,3], is_int(...))) {
    // Crash with ArgumentCountError
}
```

```
Fatal error: Uncaught ArgumentCountError: is_int() expects exactly 1 argument, 2 given
```
